### PR TITLE
Update Main with remote changes

### PR DESCRIPTION
**Pull request by wmcnicho**
Grabbing commits from `save_to_s3` allowing S3 dependency to be optional.

Works locally, testing in B9 enviornment.

@vladfi1 Let me know what you think, my modifications are contained mostly to `train.py`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -133,3 +133,6 @@ experiments/
 
 # don't track replays
 data/AllCompressed/
+
+# storage for b9 scripts
+run_scripts/
```
Comment (wmcnicho) [inline .gitignore:138]: Might be overkill, but this is where I'm storing `run-slippi.sh` on my local machine.

File: train.py [hunk 2]
```diff
@@ -78,15 +106,64 @@ def main(dataset, expt_dir, num_epochs, epoch_time, save_interval, _config, _log
...
+
+  def get_state():
+    return tf.nest.map_structure(lambda v: v.numpy(), tf_state)
+
+  def set_state(state):
+    tf.nest.map_structure(
+      lambda var, val: var.assign(val),
+      tf_state, state)
+
+  pickle_path = os.path.join(expt_dir, 'latest.pkl')
```
Comment (wmcnicho) [inline train.py:123]: My contribution starts here

**Comment by wmcnicho** [comment]
Waiting for job running at https://alpha.b-9.io/job/result.php?resultid=35

**Comment by vladfi1** [comment]
Looks good to me, thanks!

**Pull request merged**